### PR TITLE
[Context Parallel] Handle short sequence load balancing

### LINKED ISSUE #159706: Wrong shapes for context_parallel with small head size
### 🐛 Describe the bug

This testcase taken from https://docs.pytorch.org/tutorials/prototype/context_parallel.html. Run with `torchrun --standalone --nnodes=1 --nproc-per-node=2 cp_sdpa_example.py` on a 2-GPU system.
```
# file: cp_sdpa_example.py
import os

import torch
import torch.distributed as dist
import torch.nn.functional as F
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor.experimental import context_parallel
from torch.distributed.tensor.experimental._attention import context_parallel_unshard
from torch.nn.attention import sdpa_kernel, SDPBackend


def context_parallel_sdpa_example(world_size: int, rank: int):
    assert torch.cuda.is_available()
    assert dist.is_nccl_available()
    torch.cuda.set_device(f"cuda:{rank}")
    torch.cuda.manual_seed(0)

    dist.init_process_group(
        backend="nccl",
        init_method="env://",
        world_size=world_size,
        rank=rank,
    )
    device_mesh = init_device_mesh(device_type="cuda", mesh_shape=(world_size,), mesh_dim_names=("cp",))

    batch = 1
    nheads = 1
    qkv_len = 4
    dim = 4
    backend = SDPBackend.FLASH_ATTENTION
    dtype = torch.bfloat16 if backend == SDPBackend.FLASH_ATTENTION or backend == SDPBackend.CUDNN_ATTENTION else torch.float32

    qkv = [
        torch.rand(
            (batch, nheads, qkv_len, dim),
            dtype=dtype,
            requires_grad=True,
            device="cuda",
        )
        for _ in range(3)
    ]
    # specify the SDPBackend to use
    with sdpa_kernel(backend):
        out = F.scaled_dot_product_attention(*qkv, is_causal=True)

    # make a clean copy of QKV for output comparison
    cp_qkv = [t.detach().clone() for t in qkv]
    print(f"{cp_qkv[0].shape=} {type(cp_qkv[0])}")

    with sdpa_kernel(backend):
        # This `context_parallel()` performs two actions:
        # 1. Shard the tensor objects in `buffers` in-place along the dimension
        #    specified in `buffer_seq_dims`, the tensors in `buffers` and their
        #    sharding dims in `buffer_seq_dims` are organized in the same order.
        # 2. Replace the execution of `F.scaled_dot_product_attention` with a
        #    context-paralleled-enabled Ring Attention.
        with context_parallel(device_mesh, buffers=tuple(cp_qkv), buffer_seq_dims=(2, 2, 2)):
            cp_out = F.scaled_dot_product_attention(*cp_qkv, is_causal=True)
            print(f"{cp_out.shape=}")

        # The output `cp_out` is still sharded in the same way as QKV
        # the `context_parallel_unshard` API allows users to easily
        # unshard to gain the full tensor.
        (cp_out,) = context_parallel_unshard(device_mesh, [cp_out], [2])

    assert torch.allclose(
        cp_out,
        out,
        atol=(1e-08 if dtype == torch.float32 else 1e-03 * world_size),
    )


if __name__ == "__main__":
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])

    try:
        context_parallel_sdpa_example(world_size, rank)
    finally:
        dist.barrier()
        dist.destroy_process_group()
```
Observations: it'll fail if `dim` is 4 or 2. It succeeds if `dim` is 8. (Well, the `allclose` will fail, but let's ignore that for this bug report.)

Output:
```
kevin@kevin-h100-0:/mnt/clusterstorage/workspace/kevin$ torchrun --standalone --nnodes=1 --nproc-per-node=2 cp_sdpa_example.py
cp_qkv[0].shape=torch.Size([1, 1, 4, 4]) <class 'torch.Tensor'>
cp_qkv[0].shape=torch.Size([1, 1, 4, 4]) <class 'torch.Tensor'>
cp_out.shape=torch.Size([1, 1, 4, 4])
cp_out.shape=torch.Size([1, 1, 4, 4])
```
`cp_out` should be [1, 1, 2, 4].

### Versions

```
PyTorch version: 2.9.0a0+6834911
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.35

Python version: 3.10.12 (main, May 27 2025, 17:12:29) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.13-65-650-4141-22041-coreweave-amd64-85c45edc-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3

Nvidia driver version: 535.230.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           8
CPU max MHz:                        4100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5600.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hfi vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           128 MiB (64 instances)
L3 cache:                           120 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111,113,115,117,119,121,123,125,127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] clip-anytorch==2.6.0
[pip3] dctorch==0.1.2
[pip3] DISTS-pytorch==0.1
[pip3] gpytorch==1.14
[pip3] lovely-numpy==0.2.13
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.24.4
[pip3] torch==2.9.0a0+6834911
[pip3] torchaudio==2.8.0a0+b6a3368
[pip3] torchdata==0.11.0
[pip3] torchdiffeq==0.2.5
[pip3] torchsde==0.2.6
[pip3] torchtitan==0.1.0
[pip3] torchvision==0.24.0a0+f52c4f1
[pip3] triton==3.4.0+git11ec6354
[pip3] welford-torch==0.2.5
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by aditvenk**
Default context parallel load balancing now only uses the head-tail balancer when the sequence length can be split into the required head and tail chunks. Shorter sequences fall back to regular CP sharding for correctness, while preserving the previous load-balance preference after the context exits.

Fixes https://github.com/pytorch/pytorch/issues/159706

Authored by Codex.

### CODE REVIEW

**Comment by aditvenk** [comment]
@claude review the bug and the fix

**Comment by aditvenk** [comment]
@claude Examine why the new unit test is failing in CI

**Comment by aditvenk** [comment]
> Added test_context_parallel_sdpa_short_sequence to the skipped_tests list for RingAttentionTestWithLocalTensor — same treatment as the existing test_ring_attention_sdpa skip.


@claude  Will you be able to add this skip as a new commit to this branch?

File: torch/distributed/tensor/experimental/_context_parallel/_attention.py [hunk 1]
```diff
@@ -1592,13 +1594,16 @@ def context_parallel(
...
-        if original_buffer is not None:
-            buffer.resize_(original_buffer.shape)
-            buffer.copy_(original_buffer)
+    try:
+        yield
+    finally:
+        _disable_context_parallel_dispatcher_impl()
+        _cp_options.enable_load_balance = old_enable_load_balance
+
+        for buffer, original_buffer in zip(buffers, original_buffers):
```
Comment (Skylion007) [inline torch/distributed/tensor/experimental/_context_parallel/_attention.py]: Set zip kwarg to strict=True

Comment (aditvenk) [reply torch/distributed/tensor/experimental/_context_parallel/_attention.py]: Done

**Comment by aditvenk** [comment]
@claude review failing CI test and if it is the same LocalTensorMode failure from before, push a commit that disables this new unit test when under LocalTensorMode.

**Review by fegin — APPROVED** [review]

**Comment by aditvenk** [comment]
@claude Check failing CI in latest run and whether they are caused by this PR
